src/Auth0.OidcClient.Core/Tokens/IdTokenValidator.cs

```diff
@@ -0,0 +1,140 @@
+        {
+            return AssertTokenMeetsRequirements(requirements, rawIDToken, DateTime.Now);
+        }
+
+        /// <summary>
+        /// Assert that all the <see cref="IdTokenRequirements"/> are met by a JWT ID token for a given point in time.
+        /// </summary>
+        /// <param name="required"><see cref="IdTokenRequirements"/> that should be asserted.</param>
+        /// <param name="rawIDToken">Raw ID token to assert requirements against.</param>
+        /// <param name="pointInTime">The <see cref="DateTime"/> to act as "Now" in order to facilitate unit testing with static tokens.</param>
+        /// <param name="signatureVerifier">Optional <see cref="ISignatureVerifier"/> to perform signature verification and token extraction. If unspecified
+        /// <see cref="AsymmetricSignatureVerifier"/> is used against the <paramref name="required"/> Issuer.</param>
+        /// <exception cref="IdTokenValidationException">Exception thrown if <paramref name="rawIDToken"/> fails to
+        /// meet the requirements specified by <paramref name="required"/>.
+        /// </exception>
+        /// <returns><see cref="Task"/> that will complete when the token is validated.</returns>
+        internal static async Task AssertTokenMeetsRequirements(this IdTokenRequirements required, string rawIDToken, DateTime pointInTime, ISignatureVerifier signatureVerifier = null)
+        {
+            if (string.IsNullOrWhiteSpace(rawIDToken))
+            {
+                if (required.TokenIsRequired)
+                    throw new IdTokenValidationException("ID token is required but missing.");
+                return;
+            }
+
+            if (signatureVerifier == null) { 
+                var jsonWebKeys = await JsonWebKeys.GetForIssuer(required.Issuer);
+                signatureVerifier = new AsymmetricSignatureVerifier(jsonWebKeys.Keys);
+            }
+
+            var token = signatureVerifier.VerifySignature(rawIDToken);
+
+            var epochNow = EpochTime.GetIntDate(pointInTime);
+
+            // Issuer
+            if (string.IsNullOrWhiteSpace(token.Issuer))
+                throw new IdTokenValidationException("Issuer (iss) claim must be a string present in the ID token.");
+            if (token.Issuer != required.Issuer)
+                throw new IdTokenValidationException($"Issuer (iss) claim mismatch in the ID token; expected \"{required.Issuer}\", found \"{token.Issuer}\".");
+
+            // Subject
+            if (string.IsNullOrWhiteSpace(token.Subject))
+                throw new IdTokenValidationException("Subject (sub) claim must be a string present in the ID token.");
+
+            // Audience
+            var audienceCount = token.Audiences.Count();
+            if (audienceCount == 0)
+                throw new IdTokenValidationException("Audience (aud) claim must be a string or array of strings present in the ID token.");
+            if (!token.Audiences.Contains(required.Audience))
```

Comment:
`token.Audiences` will always be an array?

Reply:
Yes, the decoder always gives you an array even if there's only one item.

Reply:
Java is the same! I guess it depends on what JWT library you're using internally.